### ISSUE TITLE
False positive type mismatch on matching a slice

### ISSUE THREAD

**Issue by inflation** [rust-lang/rust-analyzer#12630]
**rust-analyzer version**: (eg. output of "Rust Analyzer: Show RA Version" command)
rust-analyzer version: 0.0.0 (434e718b6 2022-06-23)
**rustc version**: (eg. output of `rustc -V`)
1.61.0
**relevant settings**: (eg. client settings, or environment variables like `CARGO`, `RUSTUP_HOME` or `CARGO_HOME`)

```rust
let buf = [0u8; 10];
match &buf[..] {
    b"yes" => println!("yes"),
    _ => println!("no"),
}
```

<img width="395" alt="image" src="https://user-images.githubusercontent.com/2375962/175528330-4006cd62-0ad2-4035-aef7-97ff80ac8bae.png">
